README.md

```diff
@@ -93,17 +114,144 @@ run the hook from outside Vagrant).
+                              description="Standard JavaScript style and formatting linter"
+                              "(config: .eslintrc).")
+
+linter_config.do_lint()
+```
+
+Please make sure external linter (here `eslint`) is accessible via bash or in the
+virtual env where this linter will run.
 
 ## Writing custom rules
 
-TODO: Document all the features of the `RuleList` and `custom_check` system.
+You can write your own custom rules for any language using regular expression
+in zulint. Doing it is very simple and there are tons of examples available
+in [Zulip's custom_check.py file](https://github.com/zulip/zulip/blob/master/tools/linter_lib/custom_check.py).
+
+In the [above example](#adding-third-party-linters) you can add custom rules via `@linter_config.lint` decorator.
+For eg:
+
+```python
+
+from zulint.custom_rules import RuleList
+
+@linter_config.lint
+def check_custom_rules():
+    # type: () -> int
+    """Check trailing whitespace for specified files"""
+    trailing_whitespace_rule = RuleList(
+        langs=file_types,
+        rules=[{
+            'pattern': r'\s+$',
+            'strip': '\n',
+            'description': 'Fix trailing whitespace'
+        }]
+    )
+    failed = trailing_whitespace_rule.check(by_lang, verbose=args.verbose)
+    return 1 if failed else 0
+```
+
+#### RuleList
+A new custom rule is defined via the `RuleList` class. `RuleList` takes the following arguments:
+
+```python
+langs                             # The languages this rule will run on. eg: ['py', 'bash']
+rules                             # List of custom `Rule`s to run. See definition of Rule below for more details.
+max_length                        # Set a max length value for each line in the files. eg: 79
+length_exclude                    # List of files to exclude from `max_length` limit. eg: ["README"]
+shebang_rules                     # List of shebang `Rule`s to run in `langs`. Default: []
+exclude_files_in                  # Directory to exclude from all rules. eg: 'app/' Default: None
+exclude_max_length_fns            # List of file names to exclude from max_length limit. eg: [test, example] Defautl: []
+exclude_max_length_line_patterns  # List of line patterns to exclude from max_length limit. eg: ["`\{\{ api_url \}\}[^`]+`"]
+```
+
+#### Rule
+A rule is a python dictionary containing regular expression,
+which will be run on each line in the `langs`' files specified in the `RuleList`.
+It has a lot of additional features which you can use to run the pattern in
+specific areas of your codebase.
+
+Find below all the keys that a `Rule` can have along with the
+type of inputs they take.
+
+```python
+Rule = TypedDict("Rule", {
+    "bad_lines": List[str],
+    "description": str,
+    "exclude": Set[str],
+    "exclude_line": Set[Tuple[str, str]],
+    "exclude_pattern": str,
+    "good_lines": List[str],
+    "include_only": Set[str],
+    "pattern": str,
+    "strip": str,
+    "strip_rule": str,
```

Comment:
The `strip` options aren't documented?

Reply:
Yeah, Actually I think we are not using strip_rule anywhere. I will remove it. For `strip` we have this code block: 
```python
        strip_rule = rule.get('strip')  # type: Optional[str]

        ok = True
        for (i, line, line_newline_stripped, line_fully_stripped) in line_tups:
            if line_fully_stripped in exclude_lines:
                exclude_lines.remove(line_fully_stripped)
                continue
            try:
                line_to_check = line_fully_stripped
                if strip_rule is not None:
                    if strip_rule == '\n':
                        line_to_check = line_newline_stripped
                    else:
                        raise Exception("Invalid strip rule")
```
Which means we can simply make strip a boolean argument.

Reply:
or Were you thinking of adding additional rules for strip? Which I am not sure what they would be?